baybe/utils/basic.py

```diff
@@ -88,3 +89,20 @@ def group_duplicate_values(dictionary: dict[_T, _U]) -> dict[_U, list[_T]]:
     for key, value in dictionary.items():
         group.setdefault(value, []).append(key)
     return {k: v for k, v in group.items() if len(v) > 1}
+
+
+def get_init_attributes(
+    first: type, second: type, ignored_keys: tuple[str, ...] = ("self", "kwargs")
+) -> dict:
+    """Get attributes of `second` which are contained in the `__init__` of `first`.
+
+    Args:
+        first: The class whose `__init__` function is being checked.
+        second: The class whose attributes are extracted.
+        ignored_keys: A tuple of keys that should be ignored.
```

Comment:
Fully agree, this is inconsistent. Isn't `attributes` correct? I literally call `getattr` 😄 Do you have any other idea/suggestion?

Reply:
I have changed this to `get_init_arguments`, since this is what the function actually does - it gets the arguments of the `__init__` function. I tried to make the role of `filter` clear in the docstring. Please resolve if you are happy with this.
Also, feel free to use this thread for other comments on this function (resolved several that were outdated resp. not being commented on during your last reviews)

Reply:
Hi @AVHopp, first of all, thanks for the changes. I hate to be the annoying guy, but unfortunately there are still quite many issues with your function:
* The optional filter does not really work, i.e. it's not really optional. Try to run this code and you'll see what I mean:
   ```python
  import gpytorch
  
  from baybe.kernels import MaternKernel
  from baybe.utils.basic import get_init_arguments
  
  
  x = get_init_arguments(
      target=gpytorch.kernels.MaternKernel, ignored_args=("self", "kwargs")
  )
  print(x)
  
  y = get_init_arguments(
      target=gpytorch.kernels.MaternKernel,
      ignored_args=("self", "kwargs"),
      filter=MaternKernel(),
  )
  print(y)
  ```
* This one is on me: I forgot that `filter` is actually a python keyword, i.e. we shouldn't use it (surprised that no linter complained). Any alternative ideas? Maybe `reference`?
* Would be good if you could add a fully defined output type (i.e. specify the dict keys and values in the type hint)
* The way you typed/documented your function and the way you use it isn't really aligned. You speak of classes as inputs, but you actually provide an object as input to the filter argument. 
* I can live with `attributes` (though technically it should be parameters), but in the return you again talk about parameters -> still inconsistent. Either change it there or rename the function to `get_init_params` and use "parameter" everywhere
* Also, if you just read the documentation (and the function name), it is absolutely unclear what the output dictionary even contains or why it is a dictionary in the first place. I think you need to specify that in the output docstring

Reply:
Ah, I just also see that mypy failed in your pipeline run, which confirms the typing issue 👍🏼

Reply:
(I guess that make me somewhat less the bad guy since we can now just blame mypy 😄 )

Reply:
Proposed solution (not yet pushed): Would this work?
```python
def get_attributes_based_on_init_parameters(
    target_class: type,
    object: Any,
    ignored_attrs: tuple[str, ...] = ("self", "kwargs", "args"),
) -> dict[str, Any]:
    """Get attributes of an object based on the parameters of a class` ``__init__``.

    Args:
        target_class: The target class.
        object: The object whose attributes should be returned.
        ignored_attrs: Arguments that will be ignored.

    Returns:
        The dictionary of parameters.
    """
    # Get the parameters of the target signature
    params = signature(target_class.__init__).parameters
    return {p: getattr(object, p) for p in params if p not in ignored_attrs}
    ```

Reply:
Good point, will make it optional.
@AdrianSosic Could you very quickly comment on that piece of code? It seems like your confirmation (since you raised that issue) is the only remaining piece blocking this PR

Reply:
Ok, let me quickly test this. In the meantime: I don't understand @Scienfitz comment!? In the posted snippet, the parameter IS already optional (i.e. has a default, to be precise). So what exactly are we talking about?

Reply:
I guess it referes to the type hint which is still missing.

Reply:
The type hint is fine IMO, but I'll post a refined version in a minute

Reply:
That looks good to me. Will add that, see if everything works as expected and then resolve this thread. Thanks to bot of you for the feedback!

Reply:
Just one question: Why the underscore at the end of `callable_`?

Reply:
Some problems as with `filter`. `callable` is a python keyword, unfortunately. Feel free to replace this with any other appropriate name if you can find one 👍🏼